File: src/types/_Dice.ts
import * as DB from "tsondb/schema/dsl"

export const Dice = DB.TypeAlias(import.meta.url, {
  name: "Dice",
  type: () =>
    DB.Object({
      number: DB.Required({
        comment: "Number of dice of the same type. Example: 2 in 2D6.",
        type: DB.Integer({
          minimum: 1,
        }),
      }),
      sides: DB.Required({
        comment: "Number of sides on every die. Example: 6 in 2D6.",
        type: DB.IncludeIdentifier(DieType),
      }),
    }),
})

export const DieType = DB.TypeAlias(import.meta.url, {
  name: "DieType",
  comment: "Number of sides on every die. Example: 6 in 2D6.",
  type: () =>
    DB.Integer({
      minimum: 2,
    }),
})

File: Sources/OptolithDatabaseSchema/GeneratedTypes/_Dice.swift
//
//  _Dice.swift
//  OptolithDatabaseSchema
//

public struct Dice: EntitySubtype {
    /// Number of dice of the same type. Example: 2 in 2D6.
    public let number: Int

    /// Number of sides on every die. Example: 6 in 2D6.
    public let sides: DieType

    public init(number: Int, sides: DieType) {
        self.number = number
        self.sides = sides
    }
}

/// Number of sides on every dice. Example: 6 in 2D6.
public enum DieType: Int, EntitySubtype {
    case _3 = 3
    case _6 = 6
    case _20 = 20
}
